### PR TITLE
Add testing infrastructure and test cases for blpapi-node

### LINKED ISSUE #47: JS: Provide a mock JS API for testing
The [blpapi-http](http://github.com/bloomberg/blpapi-http) project has something for this currently, but will need to be adapted for the new JS API that will be addressed in #46.

### PR DESCRIPTION
**Pull request by ericlu88**
This PR adds basic testing infrastructure to blpapi-node project and an initial test suite that covers the public facing APIs.

This closes #47.

### CODE REVIEW

**Comment by ericvw** [comment]
lgtm.

File: test/blpapi.test.ts [hunk 1]
```diff
@@ -0,0 +1,616 @@
...
+                    done(new Error('Timeout did not happen!'));
+                })
+                .catch((err: Error): void => {
+                    done(err);
+                });
+        });
+    });
+
+    describe('#request()', (): void => {
+        it('should response correctly if data comes in 1 chunk', (done: Function): void => {
```
Comment (thmiceli) [inline test/blpapi.test.ts]: nit: s/response/respond

File: test/blpapi.test.ts [hunk 2]
```diff
@@ -0,0 +1,616 @@
...
+                            'HistoricalDataRequest',
+                            null,
+                            (err: Error, data?: any, isFinal?: boolean): void => {
+                                should.ifError(err);
+                                data.should.eql('FinalTestData');
+                                should(isFinal).be.true;
+                                done();
+                            });
+        });
+        it('should response correctly if data comes in 3 chunk', (done: Function): void => {
```
Comment (thmiceli) [inline test/blpapi.test.ts]: s/chunk/chunks

File: test/blpapi.test.ts [hunk 3]
```diff
@@ -0,0 +1,616 @@
...
+                            (err: Error, data?: any, isFinal?: boolean): void => {
+                                should.ifError(err);
+                                data.should.eql('FinalTestData');
+                                should(isFinal).be.true;
+                                if (++counter === 2) {
+                                    done();
+                                }
+                            });
+        });
+        it('should fail if service cannot be openned', (done: Function): void => {
```
Comment (thmiceli) [inline test/blpapi.test.ts]: s/openned/opened

File: test/blpapi.test.ts [hunk 4]
```diff
@@ -0,0 +1,616 @@
...
+                mockSession.sendTokenGenerationFailure(cid);
+            };
+            session.authenticate().catch((err: Error): void => {
+                should(err).not.be.empty;
+                err.name.should.eql('BlpApiError');
+                err.message.should.eql('Fail to generate token.');
+                done();
+            });
+        });
+        it('should fail if auth service cannot be openned', (done: Function): void => {
```
Comment (thmiceli) [inline test/blpapi.test.ts]: s/openned/opened

File: test/blpapi.test.ts [hunk 5]
```diff
@@ -0,0 +1,616 @@
...
+                should(err).not.be.empty;
+                err.name.should.eql('BlpApiError');
+                err.message.should.eql('//blp/apiauth Service Fail to Open.');
+                done();
+            });
+        });
+    });
+
+    describe('#subscribe()', (): void => {
+        it('should subscribe successfully if 1 data events occurs', (done: Function): void => {
```
Comment (thmiceli) [inline test/blpapi.test.ts]: s/events/event

File: test/blpapi.test.ts [hunk 6]
```diff
@@ -0,0 +1,616 @@
...
+                sub1 = subs[0];
+                sub2 = subs[1];
+                mockSession.sendSubscriptionData(sub1.event, sub1.cid);
+                mockSession.sendSubscriptionData(sub2.event, sub2.cid);
+            };
+            var subscription1 = new BLPAPI.Subscription('ABC', ['P']);
+            var subscription2 = new BLPAPI.Subscription('DEF', ['P']);
+            var counter: number = 0;
+            session.subscribe([ subscription1, subscription2 ]);
+            subscription1.on('data', (data: any): void => {
```
Comment (thmiceli) [inline test/blpapi.test.ts]: You don't really check that `unsubscribe` worked on `subscription2` since you don't have a listener on it.

Comment (ericlu88) [inline test/blpapi.test.ts]: good catch.

**Comment by thmiceli** [comment]
LGTM

**Comment by thmiceli** [comment]
@ericw I believe this PR can be merged now.

**Pull request merged**